src/DynamoCoreWpf/ViewModels/Search/NodeSearchElementViewModel.cs

```diff
@@ -323,35 +337,136 @@ protected virtual bool CanCreateAndConnectToPort(object parameter)
         {
             // Do not auto connect code block node since default code block node do not have output port
             if (Model.CreationName.Contains("Code Block")) return false;
+            
+            // Avoid triggering auto complete for an already connected input.
+            PortModel portModel = parameter as PortModel;
+            return parameter as PortModel != null ? portModel.CanAutoCompleteInput() : true;
+        }
+
+        private bool AutoLayoutNeeded(NodeModel originalNode, IEnumerable<NodeModel> allNodes, bool newInput, out List<NodeModel> intersectedNodes)
+        {
+            //Collect all connected input or output nodes from the original node.
+            List<NodeModel> connectedNodesToConsider = new List<NodeModel>();
+            HashSet<Guid> connectedNodesGuids = new HashSet<Guid>();
+            connectedNodesGuids.Append(originalNode.GUID);
+            var originalPorts = newInput ? originalNode.InPorts : originalNode.OutPorts;
+
+            foreach (var port in originalPorts)
+            {
+                foreach (var connector in port.Connectors)
+                {
+                    var otherNode = newInput ? connector?.Start?.Owner : connector?.End?.Owner;
+                    if (otherNode != null)
+                    {
+                        connectedNodesToConsider.Add(otherNode);
+                        connectedNodesGuids.Add(otherNode.GUID);
+                    }
+                }
+            }
+
+            //Compute a bbox of the connected nodes.
+            double minX = connectedNodesToConsider.Min(node => node.Rect.TopLeft.X);
+            double maxX = connectedNodesToConsider.Max(node => node.Rect.BottomRight.X);
+            double minY = connectedNodesToConsider.Min(node => node.Rect.TopLeft.Y);
+            double maxY = connectedNodesToConsider.Max(node => node.Rect.BottomRight.Y);
+
+            Rect2D connectedNodesBBBox = new Rect2D(minX, minY, maxX - minX, maxY - minY);
```

Comment:
maybe worth convert into utility, it could be useful later if we add some kind of visual to the whole cluster, so we could have a neat boundingbox for any node set